Add initial tranche of examples to reference manual

johnwhitington opened the pull request:
(This is work funded by the OCaml Software Foundation)

This PR introduces about eighty examples into the reference manual, which has not previously contained any.

First order principles. Examples should:

- be unobtrusive to readers who would previously have read this part of the manual
- increase the proportion of readers who find this part of the manual useful
- be substantive, not trivial
- be concise -- less than ten lines for whole `caml_example` box in output
- be idiomatic -- code one might actually write
- make use of the `toplevel` environment everywhere possible or useful
- meet the OCaml style guide
- encourage by their existence the addition of new examples when new features are added to the language

Second order principle. Examples should:

- be exhaustive -- exercise every variation of every construct in the language

This PR consists, mostly, of examples for the "Patterns" and "Expressions" sections of the chapter. The examples try to follow the list of first order principles above, but make only limited efforts with the second order principle.

There are other places in the manual where it is less obvious that examples would help, or where they might be obtrusive. And so, I would like to present the more obviously useful material first, in this PR.

The rendered sections containing new examples may be found here:

https://coherentpdf.com/tmp/lex.html
https://coherentpdf.com/tmp/patterns.html
https://coherentpdf.com/tmp/expr.html
https://coherentpdf.com/tmp/typedecl.html

Suggestions for making any of these examples more concise, more idiomatic, and more useful much appreciated. Coming up with sensible examples is hard, and outside input is very useful.

johnwhitington committed:
First tranche of examples for reference manual

Octachron commented on manual/src/refman/lex.etex:
```diff
@@ -302,16 +348,16 @@ longest first token.
 linenum-directive:
-     '#' {{"0" \ldots "9"}} '"' { string-character } '"'
+        '#' {{"0" \ldots "9"}}
+      | '#' {{"0" \ldots "9"}} '"' { string-character } '"'
```
This is probably an unintended mistake: you are reverting recent changes that fixed the line directive grammar and description.

johnwhitington replied:
Thanks. I will look through the history of these files and make sure all recent changes are taken into account.

johnwhitington replied:
Fixed.

Fourchaux commented on manual/src/refman/lex.etex:
```diff
@@ -178,11 +206,25 @@ encoding of the specified Unicode scalar value. The Unicode scalar
+let longstr =
+  "Call me Ishmael. Some years ago — never mind how long \
+  precisely — having little or no money in my purse, and \
+  nothing particular to interest me on shore, I thought I\
```
A space is missing (between I and \\) I think:
```
   nothing particular to interest me on shore, I thought I \
```

Fourchaux commented on manual/src/refman/patterns.etex:
```diff
@@ -88,6 +121,11 @@ parenthesized pattern, as in @"(" pattern_1 ":" typexpr ")"@. This
 
+\begin{caml_example}{toplevel}
+let int_triple_is_ordered ((a, b, c) : int * int * int) =
+  a < b && b < c;;
```
```
int_triple_is_ordered (1,1,2);;
- : bool = false
```
Question:  
Are two identical values not ordered  (and then a <=b && b <= c)?

johnwhitington replied:
Fixed per your suggestion. Thanks.

johnwhitington committed:
Two mistakes fixed

johnwhitington committed:
Adds changes entry

johnwhitington committed:
Re-instate linenum changes

Fourchaux commented on manual/src/refman/expr.etex:
```diff
@@ -195,6 +206,24 @@ is good style to use @'begin' \ldots 'end'@ inside control structures:
+      print_string "Not Equal: ";
+      print_int a;
+      print_string " and ";
+      print_int b
```
For a better display (and other print_* functions ...):
```
let f a b =
  if a = b then
    print_endline "Equal"
  else
    begin
      print_string "Not Equal: ";
      print_int a;
      print_string " and ";
      print_int b;
      print_newline ()
  end;;
```

xavierleroy replied:
The most popular style (as far as I know) is
```
  if ... then
    ...
  else begin
    ...
  end
```

johnwhitington replied:
Both suggestions incorporated.

Fourchaux commented on manual/src/refman/expr.etex:
```diff
@@ -460,6 +606,15 @@ true:
+  print_string (string_of_int a);
+  print_string ",";
+  print_string (string_of_int b);
+  print_string ")";;
```
For a better display, as above:  
```
  ...
  print_endline ")";;
```

Fourchaux commented on manual/src/refman/expr.etex:
```diff
@@ -676,11 +956,15 @@ has number 0; the last character has number $n-1$, where \var{n} is the
 access is out of bounds.
 
+\begin{caml_example}{toplevel}
+let iter f s =
```
The String.iter function exists in the Stdlib.
```
let iter f s =
  for i = 0 to length s - 1 do f (unsafe_get s i) done ;;
```
Isn't it better to use it?

xavierleroy replied:
If the purpose is to illustrate the `s.[i]` construct, obviously not.

xavierleroy commented on manual/src/refman/expr.etex:
```diff
@@ -273,7 +338,13 @@ matches the argument, the exception "Match_failure" is raised.
+(function (0, 0) -> "both zero"
+        | (0, _) -> "first only zero"
+        | (_, 0) -> "second only zero"
+        | _ -> "neither zero")
```
I would prefer `(_, _)` as the pattern for the last case.

xavierleroy commented on manual/src/refman/expr.etex:
```diff
@@ -285,6 +356,11 @@ This expression is equivalent to:
 
+\begin{caml_example}{toplevel}
+let p = (fun a -> fun b -> fun c -> a + b + c)
+let q = (fun a b c -> a + b + c);;
```
These are functions, so I would name them `f` and `g` rather than `p` and `q`.

xavierleroy commented on manual/src/refman/expr.etex:
```diff
@@ -321,6 +412,18 @@ expr@ is equivalent to
+   | Some false -> open_in
+   | Some true -> open_in_bin
+   | None -> open_in)
+  filename
```
A `match` that returns functions and is immediately applied afterwards is not idiomatic.  I  would write
```
  match binary with
  | Some true -> open_in_bin filename
  | Some false | None -> open_in filename
```

xavierleroy commented on manual/src/refman/expr.etex:
```diff
@@ -321,6 +412,18 @@ expr@ is equivalent to
+  filename
+
+let open_file_for_input' ?(binary=false) filename =
+  (if binary then open_in_bin else open_in) filename;;
```
Likewise.

xavierleroy commented on manual/src/refman/expr.etex:
```diff
@@ -362,6 +465,13 @@ then @expr_i@ is evaluated and its value returned as the result of the
+let rec repeat f = function
+  | n when n < 0 -> raise (Invalid_argument "repeat")
+  | 0 -> ()
+  | n -> f (); repeat f (n - 1);; 
```
What about putting the good cases first, then the error case?
```
  | 0 -> ()
  | n when n > 0 -> f (); repeat f (n - 1)
  | _ -> raise (Invalid_argument "repeat"
```

xavierleroy commented on manual/src/refman/expr.etex:
```diff
@@ -468,9 +623,21 @@ the value of @expr_2@ if @expr_1@ evaluates to the boolean @"true"@,
+let debug = ref false
+
+let log msg =
+  if !debug then prerr_string (msg ^ "\n");;
```
Or just `prerr_endline msg`

xavierleroy commented on manual/src/refman/expr.etex:
```diff
@@ -487,11 +654,28 @@ value of the whole @'match'@ expression. The evaluation of
+let rec sum l =
+  match l with
+  | [] -> 0
+  | h::t -> h + sum t;; 
```
Here and elsewhere I'd prefer spaces around `::`, as in `h :: t`.

xavierleroy commented on manual/src/refman/expr.etex:
```diff
@@ -668,6 +938,16 @@ the array denoted by @expr_1@, replacing element number @expr_2@ by
 the access is out of bounds. The value of the whole expression is @'()'@.
 
+\begin{caml_example}{toplevel}
+let mul arr n =
```
Maybe `scale` is a better name for this function than `mul`?  As in "scale this array by `n`".

xavierleroy commented on manual/src/refman/expr.etex:
```diff
@@ -668,6 +938,16 @@ the array denoted by @expr_1@, replacing element number @expr_2@ by
+  done
+
+let x = [|1; 10; 100|]
+let _ = mul x 2;;
```
Would you like to print `x` afterwards?

johnwhitington replied:
Ah, it does. Because by the time the output is printed, the mutation has already happened:
```
# let scale arr n =
    for x = 0 to Array.length arr - 1 do
      arr.(x) <- arr.(x) * n
    done

  let x = [|1; 10; 100|]
  let _ = scale x 2;;
val scale : int array -> int -> unit = <fun>
val x : int array = [|2; 20; 200|]
```

Octachron replied:
Note that you can add `;;` after the definition of `x`  and then show the value of `x` after calling `scale`.

xavierleroy commented on manual/src/refman/lex.etex:
```diff
@@ -116,6 +134,12 @@ It is written in decimal and interpreted as a power of 2.
 (and ignored) within floating-point literals.
 
+\begin{caml_example}{toplevel}
+let pi = 3.14159265358979312
```
Maybe add some `_` to improve legibility?  `3.141_592_653_589_793_12`

xavierleroy commented on manual/src/refman/lex.etex:
```diff
@@ -116,6 +134,12 @@ It is written in decimal and interpreted as a power of 2.
+\begin{caml_example}{toplevel}
+let pi = 3.14159265358979312
+let small_negative = -1e-5
+let hex = 0x1.fp3;;
```
I like `let machine_epsilon = 0x1p-52;; (* 2 to the -52 *)`

xavierleroy commented on manual/src/refman/lex.etex:
```diff
@@ -178,11 +206,25 @@ encoding of the specified Unicode scalar value. The Unicode scalar
+let longstr =
+  "Call me Ishmael. Some years ago — never mind how long \
+  precisely — having little or no money in my purse, and \
+  nothing particular to interest me on shore, I thought I \
+  would sail about a little and see the watery part of t\
```
You could also show how to put whitespace at the beginning of a line:
```suggestion
  nothing particular to interest me on shore, I thought I\
  \ would sail about a little and see the watery part of t\
```

xavierleroy commented on manual/src/refman/expr.etex:
```diff
@@ -570,6 +786,13 @@ selected. If none of the patterns matches the value of @expr@, the
 exception value is raised again, thereby transparently ``passing
 through'' the @'try'@ construct.
 
+\begin{caml_example}{toplevel}
+let option_exception f x =
+  try Some (f x) with
+  | Assert_failure _ as e -> raise e
+  | _ -> None;;
+\end{caml_example}
```
Looks like you're reusing an example from the "Patterns" section.  This example does not illustrate the automatic "reraise"...

johnwhitington replied:
So if I changed it to:

`Assert_failure (a, b, c) -> raise (Assert_failure (a, b, c))`

It would be detected as a reraise?

xavierleroy replied:
Sorry if I was unclear.  The "automatic reraise" behavior I'm alluding to is the fact that
```
  try e1 with Not_found -> e2
```
is equivalent to
```
  try e1 with Not_found -> e2 | exn -> raise exn
```
Surely you want to illustrate this fact?

Moreover, exception handlers with catch-all cases that don't re-raise the exception are frowned upon.  In your example, if function `f` raises `Invalid_argument` or `Out_of_memory` or `Sys.break`, all these exceptions are caught and turned into a non-informative `None` result.  Don't do that, and don't use that as an example.

xavierleroy commented on manual/src/refman/patterns.etex:
```diff
@@ -80,6 +106,13 @@ The pattern @pattern_1 "as" value-name@ matches the same values as
 the name @value-name@ is bound to the matched value, in addition to the
 bindings performed by the matching against @pattern_1@.
 
+\begin{caml_example}{toplevel}
+let option_exception f x =
+  try Some (f x) with
+  | Assert_failure _ as e -> raise e
+  | _ -> None;;
+\end{caml_example}
```
As discussed over email, I would suggest the following, simpler example for `as` patterns:
```
let sort_pair ((x, y) as p) = if x <= y then p else (y, x)
```

xavierleroy reviewed (commented):
Lots of good examples here.  Some comments and suggestions below.

dra27 commented:
As it happens, I first brought the rendered pages up on my iPhone, where the CSS apparently limits the samples to 40 characters' width. The majority of these samples actually fit within that, so perhaps for some of the longer ones (e.g. the multiple line comment) it might be worth breaking them for 40 characters or is there even the ability to use something like `<wbr>` to hint at an extra break for small devices? Similarly, `let quoted_greeting ={|"Hello, World!|}` still conveys the meaning, but doesn't require side-scrolling to get to the point of the example.

By virtue of being a fixed-width font, the text of the examples look bigger relative to the main body. I wonder for the examples if having the font a point smaller would look neater, but my eyesight is strange (a smaller font would also slightly increase the number of characters).

Also pursuant to smaller screens, and possibly "clutter" in general. I like very much the use of the toplevel (we've done similar with Jupyter notebooks for teaching OCaml at Cambridge). However, in many cases the output largely repeats the code (e.g. the underscores in integer literals example). Is there some CSS wizardry which can be done at this stage to collapse the toplevel replies an expandable `...`?

johnwhitington commented:
> As it happens, I first brought the rendered pages up on my iPhone, where the CSS apparently limits the samples to 40 characters' width. The majority of these samples actually fit within that, so perhaps for some of the longer ones (e.g. the multiple line comment) it might be worth breaking them for 40 characters or is there even the ability to use something like `<wbr>` to hint at an extra break for small devices? Similarly, `let quoted_greeting ={|"Hello, World!|}` still conveys the meaning, but doesn't require side-scrolling to get to the point of the example.

On Safari on iPhone, I get a 33-char-wide box which scrolls horizontally, rather than wrapping. Does yours actually wrap? Wrapping would be really bad here I think.

> Also pursuant to smaller screens, and possibly "clutter" in general. I like very much the use of the toplevel (we've done similar with Jupyter notebooks for teaching OCaml at Cambridge). However, in many cases the output largely repeats the code (e.g. the underscores in integer literals example). Is there some CSS wizardry which can be done at this stage to collapse the toplevel replies an expandable `...`?

In the example of the underscores, the fact that OCaml prints the number back without such underscores is perhaps interesting for the reader. I always want to show the types, because I think it's important. Now, in examples which are just type definitions, with no values, we could indeed use `{caml_example}{verbatim}` instead of `{caml_example}{toplevel}` to reduce the clutter of duplication.

johnwhitington committed:
Responses to first round of review comments

johnwhitington commented:
Fixed per comments of @Fourchaux, @xavierleroy, @dra27. If I made no comment in response to one of yours, that means I accepted the suggested fix.

Updated rendered version at https://coherentpdf.com/tmp/index.html -- you may need to reload each page to circumvent browser caching.

gasche commented on manual/src/refman/patterns.etex:
```diff
@@ -59,16 +59,42 @@ A pattern that consists in a value name matches any value,
+\begin{caml_example}{toplevel}
+let print_int_opt = function
+  | Some x -> print_int x
+  | _ -> ();;
```
If someone were to write this code and send it for review, I would ask for `_` to be turned into `None` to be more precise and non-fragile.

Here would be an example where I would not suggest removing the underscore:

```ocaml
let is_empty = function
| [] -> true
| _::_ -> false
```

gasche commented on manual/src/refman/patterns.etex:
```diff
@@ -101,6 +137,14 @@ in case some value~$v$ matches @pattern_1 "|" pattern_2@, the bindings
 
+\begin{caml_example}{toplevel}
+type shape =
+  Square of float | Rect of (float * float) | Circle of float
```
Nitpick, but could it be `| Square of float`? This looks much better in the multiline version, and even here it would make it consistent with the matching below.

gasche commented on manual/src/refman/patterns.etex:
```diff
@@ -101,6 +137,14 @@ in case some value~$v$ matches @pattern_1 "|" pattern_2@, the bindings
+
+let is_rectangular = function
+  | Square _ | Rect _ -> true
+  | _ -> false;;
```
`Circle _ -> false`?

gasche commented on manual/src/refman/patterns.etex:
```diff
@@ -135,13 +204,26 @@ is a shorthand for the following or-pattern:
+
+let to_string = function
+  | #any_rectangle -> "Rectangle"
+  | `Circle _ -> "Circle";;
```
This example is fine, but #-patterns are typically coupled with `as` to refine the type of the input, and I think it would be nice to examplify it here:

```ocaml
type 'a rectangle = [ `Square of 'a | `Rectangle of 'a * 'a ]
type 'a shape = [ `Circle of 'a | 'a rectangle ]

let try_rectangle : 'a shape -> 'a rectangle option = function
  | #rectangle as r -> Some r
  | `Circle _ -> None
```

johnwhitington replied:
Is the type annotation required in this example to get the behaviour you desire? If so, i'd rather not use it. If not, I'm happy to.

gasche replied:
It's not required, but it helps figuring out what happens and what we are trying to do. Maybe the toplevel signature output also does that, and then it is okay to drop the annotations. (In generally when I write code using polymorphic variants, I *do* annotate carefully, because in my experience it catches my mistakes faster and more locally.)

gasche commented on manual/src/refman/patterns.etex:
```diff
@@ -199,6 +308,12 @@ Matching a value with a @pattern-matching@ where some patterns
+let lazy_option_map f = function
+  | lazy (Some x) -> Some (Lazy.force f x)
+  | _ -> None;;
+\end{caml_example}
```
Actually this implementation is "wrong", it forces the thunk on application where there is no need for mapping.

You want to write:
```ocaml
let lazy_option_map f lo =
  lazy (match lo with
    | lazy (Some x) -> let (lazy y) = f x in Some y
    | lazy None -> None)
```

Note: I think `Lazy.force f x` was meant to be `Lazy.force (f x)`, as the current one makes the type not look like a map.

I think that there are simpler examples for functions that force their lazy inputs, instead of trying to preserve laziness.

```ocaml
let force_opt : int lazy option -> int = function
| Some (lazy n) -> n
| None -> 0
```

johnwhitington replied:
The example is taken (modified only slightly) from the Lazy module's documentation!

https://caml.inria.fr/pub/docs/manual-ocaml/libref/Lazy.html

Can you suggest an example suitable both for the Patterns chapter, and to fix the Lazy documentation (I've never used Lazy)?

gasche commented on manual/src/refman/patterns.etex:
```diff
@@ -229,6 +344,13 @@ A pattern match must contain at least one value case. It is an error if
+  match (List.hd l, List.hd (List.rev l)) with
+  | (f, l) -> Some (l - f)
+  | exception (Failure _) -> None;;
+\end{caml_example}
```
I would complain if someone wrote this code without pattern-matching on the empty list first. Maybe:

```ocaml
let find_opt p li =
  match List.find p li with
  | exception Not_found -> None
  | v -> Some v
```

(The general theme being of calling a function that is known to return exceptions in some "standard" failure scenarios that cannot be easily discarded by cheap input verification. Yes, in many such cases we nowadays provide the `_opt` version itself in the standard library.)

gasche reviewed (commented):
Nice work! I looked at the "patterns" section and have some nitpicking.

johnwhitington committed:
Address some of @gasche's pattern suggestions

johnwhitington committed:
machine epsilon, missed this one earlier

johnwhitington committed:
Final fixes for first round of reviews

johnwhitington commented:
All fixes to first round of reviews now done.

Updated rendered version at https://coherentpdf.com/tmp/index.html -- you may need to reload each page to circumvent browser caching.

gasche commented:
Glancing at the expression examples, there are a couple cases where I think putting two sentences after a single toplevel prompt hurts readability somewhat.

```
# let gen () = let exception A in A
  let () = assert(gen () = gen ());;
Exception: Assert_failure ("expr.etex", 2, 9).
```

I assume that it is a limitation of the current `caml_example` environment that we cannot have one prompt for each sentence? @Octachron, how much work would it be to implement explicit macros to have a sequence of prompts within the same code block?

(Note: in the meantime adding a ;; after each sentence might help, or a blank line as you used in some cases.)

Octachron commented:
What would be the difference with using `;;` to mark group of sentences to be printed independently?

johnwhitington commented:
It would be nice to be able to produce the following:

```
# let x = 1;;
val x : int = 1
# let y = x + 4;;
val y : int = 5
```

At the moment, as far as I am aware, we can only produce:

```
# let x = 1
  let y = x + 4;;
val x : int = 1
val y : int = 5
```

johnwhitington committed:
Spacing

johnwhitington committed:
Update reviewers

johnwhitington commented:
All points raised have been addressed, so this can be merged unless there are any other suggestions.

Octachron commented on manual/src/refman/expr.etex:
```diff
@@ -809,6 +1120,19 @@ with a closed object type.
+  val mutable l = l
+  method next =
+    if l = [] then raise (Failure "empty list");
+    let r = List.hd l in l <- List.tl l; r
```
Is there a reason to not use a normal pattern matching here?

johnwhitington replied:
Thanks. Fixed in all examples. I also changed the parameter to `lst` so we have `mutable l = lst` not `mutable l = l` which is perhaps confusing.

johnwhitington committed:
Fix oo examples

Octachron commented:
@johnwhitington , concerning the examples using:

```ocaml
\begin{caml_example}{toplevel}
let x = 1;;
let y = 2;;
\end{caml_example}
```
produces
>```
># let x = 1;;
>val x : int = 1
>```

>```
># let y = 2;;
>val x : int = 2
>```

in two separated divs. Would you want to have the output merged in one div?

johnwhitington commented:
Yes, it would be great to have the option to have them in one div or not i.e

```
\begin{caml_example}[singlediv]{toplevel}
let x = 1;;
let y = 1;;
\end{caml_example}
```

Octachron commented:
This seems ready to be merged to me.
@gasche , @xavierleroy: do you have another round of review planned?

xavierleroy reviewed (approved):
I had a quick second look, and it all looks good to me.  @Octachron, you can merge when you're happy too.

avsm commented on manual/src/refman/patterns.etex:
```diff
@@ -121,12 +172,29 @@ of length $n$ whose elements match @pattern_1@ \ldots @pattern_n@,
+      let ss, gs = split t in
+        match h with
+        | `Sheep _ as s -> (s :: ss, gs)
+        | `Goat _ as g -> (ss, g :: gs);;
```
I can never decide if this is good style or not, but the brackets around the tuples here aren't necessary:

```
let rec split = function
   | [] -> [], []
   | h :: t ->
       let ss, gs = split t in
         match h with
         | `Sheep _ as s -> s :: ss, gs
         | `Goat _ as g -> ss, g :: gs ;;
```

(this matches the lack of brackets on the pattern matching end over at `let ss, sg = ...`

johnwhitington replied:
Per the OCaml style guide https://ocaml.org/learn/tutorials/guidelines.html (and I agree - well, mostly. I can't decide either):

How to write pairs

A tuple is parenthesized and the commas therein (delimiters) are each followed by a space: (1, 2), let triplet = (x, y, z)...

Commonly accepted exceptions:

Definition of the components of a pair: In place of let (x, y) = ..., you can write let x, y = ....
Justification: The point is to define several values simultaneously, not to construct a tuple. Moreover, the pattern is set off nicely between let and =.

Matching several values simultaneously: It's okay to omit parentheses around n-tuples when matching several values simultaneously.

match x, y with
| 1, _ -> ...
| x, 1 -> ...
| x, y -> ...

Justification: The point is to match several values in parallel, not to construct a tuple. Moreover, the expressions being matched are set off by match and with, while the patterns are set off nicely by | and ->.

avsm replied:
I defer to the superior Chicago Manual of OCaml Style :-)

Octachron commented:
I see some unicode error when trying to build the manual locally. @johnwhitington , do you want me to push all the unicode minutiae (declaring the characters, adding it to the listing literates, etc...) directy?

johnwhitington commented:
Yes, please! I was unaware anything special would be required -- since I only looked at the HTML output in detail -- I just made sure the other kinds of manual built.

Octachron committed:
manual: unicode character declarations

Octachron commented:
Done. If you are fine with the pushed changes, I will merge.

johnwhitington commented:
Thanks. To the extent that I understand it, it looks fine!
